pkg/orchestrator/assetscanestimationwatcher/watcher.go

```diff
@@ -0,0 +1,269 @@
+		assetScanEstimationID, ok := assetScanEstimation.GetID()
+		if !ok {
+			logger.Warnf("Skipping due to invalid AssetScanEstimation: ID is nil: %v", assetScanEstimation)
+			continue
+		}
+		scanEstimationID, ok := assetScanEstimation.GetScanEstimationID()
+		if !ok {
+			logger.Warnf("Skipping due to invalid AssetScanEstimation: ScanEstimation.ID is nil: %v", assetScanEstimation)
+			continue
+		}
+		assetID, ok := assetScanEstimation.GetAssetID()
+		if !ok {
+			logger.Warnf("Skipping due to invalid AssetScanEstimation: Asset.ID is nil: %v", assetScanEstimation)
+			continue
+		}
+
+		events = append(events, AssetScanEstimationReconcileEvent{
+			AssetScanEstimationID: assetScanEstimationID,
+			ScanEstimationID:      scanEstimationID,
+			AssetID:               assetID,
+		})
+	}
+
+	return events, nil
+}
+
+// nolint:cyclop
+func (w *Watcher) Reconcile(ctx context.Context, event AssetScanEstimationReconcileEvent) error {
+	logger := log.GetLoggerFromContextOrDiscard(ctx).WithFields(event.ToFields())
+	ctx = log.SetLoggerForContext(ctx, logger)
+
+	assetScanEstimation, err := w.backend.GetAssetScanEstimation(ctx, event.AssetScanEstimationID, models.GetAssetScanEstimationsAssetScanEstimationIDParams{
+		Expand: utils.PointerTo("asset"),
+	})
+	if err != nil {
+		return fmt.Errorf("failed to get AssetScanEstimation with %s id: %w", event.AssetScanEstimationID, err)
+	}
+
+	state, ok := assetScanEstimation.GetState()
+	if !ok {
+		return fmt.Errorf("cannot determine state of AssetScanEstimation with %s id", event.AssetScanEstimationID)
+	}
+
+	logger.Tracef("Reconciling AssetScanEstimation state: %s", state)
+
+	switch state {
+	case models.AssetScanEstimationStateStatePending:
+		if err = w.reconcilePending(ctx, &assetScanEstimation); err != nil {
+			return err
+		}
+	case models.AssetScanEstimationStateStateAborted:
+		if err = w.reconcileAborted(ctx, &assetScanEstimation); err != nil {
+			return err
+		}
+	default:
+	}
+
+	return nil
+}
+
+// nolint:cyclop
+func (w *Watcher) reconcilePending(ctx context.Context, assetScanEstimation *models.AssetScanEstimation) error {
+	logger := log.GetLoggerFromContextOrDiscard(ctx)
+
+	logger.Debugf("reconciling pending asset scan estimations")
+
+	assetScanEstimationID, ok := assetScanEstimation.GetID()
+	if !ok {
+		return errors.New("invalid AssetScanEstimation: ID is nil")
+	}
+
+	if assetScanEstimation.Asset == nil || assetScanEstimation.Asset.AssetInfo == nil {
+		return errors.New("invalid AssetScanEstimation: Asset or AssetInfo is nil")
+	}
+
+	asset := &models.Asset{
+		AssetInfo:    assetScanEstimation.Asset.AssetInfo,
+		FirstSeen:    assetScanEstimation.Asset.FirstSeen,
+		Id:           &assetScanEstimation.Asset.Id,
+		LastSeen:     assetScanEstimation.Asset.LastSeen,
+		Revision:     assetScanEstimation.Asset.Revision,
+		ScansCount:   assetScanEstimation.Asset.ScansCount,
+		Summary:      assetScanEstimation.Asset.Summary,
+		TerminatedOn: assetScanEstimation.Asset.TerminatedOn,
+	}
```

Comment:
Why are we creating a new models.Asset here when we have assetScanEstimation.Asset $expanded? I think we can just change getLatestAssetScanStats to receive a AssetScanRelationship object instead.

Reply:
We need Asset also for Estimate(), do you think we should change it there as well to receive a AssetScanRelationship object?

Reply:
Hmm thats a great point, I assume we can't cleanly cast between Asset and AssetRelationship? Perhaps AssetRelationship needs a convert helper function `ToAsset()` or something. We can use this library
https://github.com/anchore/go-struct-converter to remove the need to manually map all the fields in the struct.

Reply:
Yes, we need some general convert func between these two, in order not to manually convert every field. Maybe we can also use json marshal/unmarshal